File: android/app/src/main/java/com/izzatismail/midtrim/domain/usecase/MergeVideoSegmentsUseCase.kt
package com.izzatismail.midtrim.domain.usecase

import com.izzatismail.midtrim.domain.entity.ExportQuality
import com.izzatismail.midtrim.domain.error.TrimError
import com.izzatismail.midtrim.domain.repository.VideoTrimmer

class MergeVideoSegmentsUseCase(
    private val trimmer: VideoTrimmer,
    private val qualityResolver: ResolveExportQualityUseCase
) {
    suspend fun execute(
        segmentUris: List<String>,
        isPaidUser: Boolean,
        sourceWidth: Int,
        sourceHeight: Int
    ): String {
        if (segmentUris.isEmpty()) throw TrimError.MergeFailed("No segments to merge.")

        val quality: ExportQuality = qualityResolver.execute(isPaidUser, sourceWidth, sourceHeight)
        return trimmer.merge(segmentUris, quality)
    }
}

File: ios/MidTrim/MidTrim/Domain/UseCases/MergeVideoSegmentsUseCase.swift
import Foundation

struct MergeVideoSegmentsUseCase {
    private let trimmer: VideoTrimmerProtocol
    private let qualityResolver: ResolveExportQualityUseCase

    init(trimmer: VideoTrimmerProtocol, qualityResolver: ResolveExportQualityUseCase) {
        self.trimmer = trimmer
        self.qualityResolver = qualityResolver
    }

    func execute(segmentURIs: [String], isPaidUser: Bool, sourceResolution: CGSize) async throws -> String {
        guard !segmentURIs.isEmpty else {
            throw TrimError.mergeFailed("No segments to merge.")
        }

        let quality = qualityResolver.execute(isPaidUser: isPaidUser, sourceResolution: sourceResolution)
        return try await trimmer.merge(segmentURIs: segmentURIs, outputQuality: quality)
    }
}
